src/playground.rs

```diff
@@ -0,0 +1,275 @@
+    #[serde(rename = "crateType")]
+    crate_type: CrateType,
+    mode: Mode,
+    tests: bool,
+}
+
+impl PlaygroundCode {
+    fn new(code: String) -> Self {
+        PlaygroundCode {
+            channel: Channel::Nightly,
+            edition: Edition::E2018,
+            code,
+            crate_type: CrateType::Binary,
+            mode: Mode::Debug,
+            tests: false,
+        }
+    }
+
+    fn url_from_gist(&self, gist: &str) -> String {
+        let version = match self.channel {
+            Channel::Nightly => "nightly",
+            Channel::Beta => "beta",
+            Channel::Stable => "stable",
+        };
+
+        let edition = match self.edition {
+            Edition::E2015 => "2015",
+            Edition::E2018 => "2018",
+        };
+
+        let mode = match self.mode {
+            Mode::Debug => "debug",
+            Mode::Release => "release",
+        };
+
+        format!(
+            "https://play.rust-lang.org/?version={}&mode={}&edition={}&gist={}",
+            version, mode, edition, gist
+        )
+    }
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "snake_case")]
+enum Channel {
+    Stable,
+    Beta,
+    Nightly,
+}
+
+impl FromStr for Channel {
+    type Err = Box<dyn std::error::Error>;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "stable" => Ok(Channel::Stable),
+            "beta" => Ok(Channel::Beta),
+            "nightly" => Ok(Channel::Nightly),
+            _ => Err(format!("invalid release channel `{}`", s).into()),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+enum Edition {
+    #[serde(rename = "2015")]
+    E2015,
+    #[serde(rename = "2018")]
+    E2018,
+}
+
+impl FromStr for Edition {
+    type Err = Box<dyn std::error::Error>;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "2015" => Ok(Edition::E2015),
+            "2018" => Ok(Edition::E2018),
+            _ => Err(format!("invalid edition `{}`", s).into()),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+enum CrateType {
+    #[serde(rename = "bin")]
+    Binary,
+    #[serde(rename = "lib")]
+    Library,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "snake_case")]
+enum Mode {
+    Debug,
+    Release,
+}
+
+impl FromStr for Mode {
+    type Err = Box<dyn std::error::Error>;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s {
+            "debug" => Ok(Mode::Debug),
+            "release" => Ok(Mode::Release),
+            _ => Err(format!("invalid compilation mode `{}`", s).into()),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct PlayResult {
+    success: bool,
+    stdout: String,
+    stderr: String,
+}
+
+fn run_code(args: &Args, code: &str) -> Result<String> {
+    let mut errs = String::new();
+
+    let channel = args.params.get("channel").unwrap_or(&"nightly");
+    let mode = args.params.get("mode").unwrap_or(&"debug");
+    let edition = args.params.get("edition").unwrap_or(&"2018");
```

Comment:
Can we use `unwrap_or_else` here instead? No point in doing extra evaluation if its not needed.